# Notebook #1: Python, Pandas, and Movies!
**Name: Nick Peters

**Due 9/13 before class**

The goal of this notebook is to become comfortable with using `python` and `pandas` as well as grabbing specific slices of the dataset. 

Start off by importing the data. Make sure the data is in your Google Drive, and that you update the path to point to where you have your data saved in Drive.

In [ ]:
#selecting columns in dataframes
import pandas
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#import the data:
#make sure the path on the line below corresponds to the path where you put your dataset.
movies = pandas.read_csv('/content/drive/MyDrive/CS167 Datasets/MoviesData.csv')
movies

Title  Year  Age  IMDb  Rotten Tomatoes  \
0                           Inception  2010  13+   8.8             87.0   
1                          The Matrix  1999  18+   8.7             87.0   
2              Avengers: Infinity War  2018  13+   8.5             84.0   
3                  Back to the Future  1985   7+   8.5             96.0   
4      The Good, the Bad and the Ugly  1966  18+   8.8             97.0   
...                               ...   ...  ...   ...              ...   
16692       The Ghosts of Buxley Hall  1980  NaN   6.2              NaN   
16693                  The Poof Point  2001   7+   4.7              NaN   
16694           Sharks of Lost Island  2013  NaN   5.7              NaN   
16695              Man Among Cheetahs  2017  NaN   6.6              NaN   
16696                In Beaver Valley  1950  NaN   NaN              NaN   

                            Directors        Genre  Runtime        Country  
0                   Christopher Nolan       Action    148.0  United States  
1      Lana Wachowski,Lilly Wachowski       Action    136.0  United States  
2             Anthony Russo,Joe Russo       Action    149.0  United States  
3                     Robert Zemeckis    Adventure    116.0  United States  
4                        Sergio Leone      Western    161.0          Italy  
...                               ...          ...      ...            ...  
16692                    Bruce Bilson       Comedy    120.0  United States  
16693                     Neal Israel       Comedy     90.0  United States  
16694                    Neil Gelinas  Documentary      NaN  United States  
16695            Richard Slater-Jones  Documentary      NaN  United States  
16696                     James Algar  Documentary     32.0  United States  

[16697 rows x 9 columns]

In [ ]:
print(movies.groupby(['Year'])["Genre"].apply(lambda x: x.value_counts().head(2)).to_string())

Year             
1902  Short            1
1912  Biography        1
1913  Comedy           1
      Short            1
1915  Drama            2
      Comedy           2
1916  Drama            1
1917  Adventure        1
1918  Drama            1
      Action           1
1919  Drama            2
1920  Fantasy          2
      Drama            1
1921  Drama            2
      Action           1
1922  Short            2
      Drama            1
1923  Comedy           2
      Animation        1
1924  Comedy           1
1925  Drama            3
      Comedy           1
1926  Action           1
      Drama            1
1927  Comedy           1
      Adventure        1
1928  Action           1
      Comedy           1
1929  Comedy           2
      Crime            2
1930  Comedy           5
      Biography        2
1931  Comedy           7
      Drama            5
1932  Action           7
      Comedy           6
1933  Drama           10
      Action           9
1934  Action          13
      D

## Exercise #1: 
Print out all of the possible values for the Genres.
- *Hint: I'm looking for each possible Genre to be ouput once. For example, if I tossed a coin 3 times and got Heads, Tails, Heads, the possible values of the coin flip are Heads and Tails. There is a function for this. It should require one line of code.*

In [ ]:
print(movies['Genre'].unique())

['Action' 'Adventure' 'Western' 'Animation' 'Biography' 'Drama' 'Crime'
 'Comedy' 'Documentary' 'Family' 'Horror' 'Thriller' 'Mystery' nan
 'Fantasy' 'Romance' 'Short' 'Sci-Fi' 'Talk-Show' 'Sport' 'Reality-TV'
 'Musical' 'War' 'News' 'History' 'Film-Noir']


## Exercise #2:
Create a subset of data representing movies from the year 2000-2010 (including both 2000 and 2010) with the value "Action" in the Genre column. Save this DataFrame to the variable name `older_action_movies`. Display the first 5 lines of this subset.

In [ ]:
older_action_movies = movies[(movies['Year'] > 1999) & (movies['Year'] < 2011) & (movies['Genre'] == "Action")]
older_action_movies[:5]

Title  Year  Age  IMDb  Rotten Tomatoes  \
0                     Inception  2010  13+   8.8             87.0   
27                   District 9  2009  18+   7.9             90.0   
36              Minority Report  2002  13+   7.6             90.0   
43                       Ip Man  2008  18+   8.0             85.0   
48  Scott Pilgrim vs. the World  2010  13+   7.5             81.0   

            Directors   Genre  Runtime        Country  
0   Christopher Nolan  Action    148.0  United States  
27     Neill Blomkamp  Action    112.0   South Africa  
36   Steven Spielberg  Action    145.0  United States  
43         Wilson Yip  Action    106.0      Hong Kong  
48       Edgar Wright  Action    112.0  United States

## Exercise #3
Using `older_action_movies`, what is the *median* **IMDb** rating?

In [ ]:
print(older_action_movies['IMDb'].median())

5.0


## Exercise #4
Create a new subset with only movies with `Rotten Tomatoes` values < 20 and `IMDb` values < 3.0 (these should be awful movies!). Include only 4 columns: the `Title`, `Year`, `IMDb`, and `Rotten Tomatoes`. Display all of the rows in this subset.


In [ ]:
rotten_movies = movies[ (movies['Rotten Tomatoes'] < 20) & (movies['IMDb'] < 3.0)]
rotten_movies = rotten_movies[['Title', 'Year', 'IMDb', 'Rotten Tomatoes']]
rotten_movies

Title  Year  IMDb  Rotten Tomatoes
3197                 Tees Maar Khan  2010   2.5             11.0
8466    The Haunting of Sharon Tate  2019   2.8             19.0
9213                     BloodRayne  2005   2.9              4.0
10377  Mega Shark vs. Giant Octopus  2009   2.5             18.0
10463                        Race 3  2018   2.0              9.0
11067    Birdemic: Shock and Terror  2010   1.8             19.0
11707          The Amityville Curse  1990   2.9             17.0
12397                 Clowntergeist  2017   2.9             17.0
13548                  Mega Piranha  2010   2.4              8.0

## Exercise #5

What is the average (mean) IMDb score for each of the different Genres?  (I'm looking at the average of the 'IMDb' column, grouped by Genre--so Action v Adventure v Animation etc.). Look at this list to determine which genre has the highest average.
- In addition to supplying the code to output the averages for each Genre,answer the question of the hightest average in a text cell ...*

In [ ]:
#['Action' 'Adventure' 'Western' 'Animation' 'Biography' 'Drama' 'Crime' 'Comedy' 'Documentary' 'Family' 'Horror' 'Thriller' 'Mystery' nan
#'Fantasy' 'Romance' 'Short' 'Sci-Fi' 'Talk-Show' 'Sport' 'Reality-TV'
#'Musical' 'War' 'News' 'History' 'Film-Noir']
movies_copy = movies.copy()
print(movies_copy.size)
movies_copy = movies_copy.dropna()
print(movies_copy.size)
for i in movies_copy['Genre'].unique():
  specific_genre = movies[movies['Genre'] == i]
  print(i +":",specific_genre['IMDb'].mean())


150273
29790
Action: 5.375986975986976
Adventure: 5.687123745819398
Western: 5.715
Animation: 6.23609022556391
Biography: 6.588780487804878
Drama: 6.042693143034439
Crime: 5.8660170523751525
Comedy: 5.876014536644457
Documentary: 6.943205721497686
Family: 5.529746835443039
Horror: 4.557747747747748
Mystery: 5.529473684210526
Romance: 5.7749999999999995
Thriller: 5.126696832579186
Fantasy: 4.689333333333332
Short: 6.3538461538461535
Sci-Fi: 4.217283950617284
Sport: 6.718181818181818
War: 5.35
Musical: 6.251724137931035


##Answer to Question in Exercise #5
From the smaller list of movies that did not have nan values, I found that movies in the sport genre had the highest average, sitting at 6.718181818181818.

## Exercise #6

Do something cool. Using the movies dataset, get creative and show me something about the dataset. Use a text cell to explain what you did.

##Explanation
Using a double nested loop, the below program runs through each unique year that movies were published. For each year, it runs through each movie in the dataset. For the unique year it is looking at, it counts the number of movies of each genre. When the inner loop gets to a year that is beyond the unique year, the inner loop is broken. When the program finishes, you are left with a large printed list of years and the genre of movie that was most common in that year. With a large dataset, of course there are some years where the max genre is the same for mutiple genres.

In [ ]:
import numpy as np
movies_by_year = movies.sort_values(by=['Year'])
movies_by_year = movies_by_year.reset_index(drop=True)
list_of_genres = movies['Genre'].unique()

for i in movies_by_year['Year'].unique():
  genre_count = np.zeros(list_of_genres.size)
  for x in range (movies_by_year['Year'].size):
    if (movies_by_year['Year'].iloc[x] > i):
      break
    #print(movies_by_year['Genre'].iloc[x])
    if (movies_by_year['Year'].iloc[x] == i):
      genre_count[np.where(list_of_genres == movies_by_year['Genre'].iloc[x])] += 1
  
  highest_genre = genre_count.max()
  print('Year:', i)
  print('Most movies produced in a genre:', highest_genre, list_of_genres[np.where(genre_count == highest_genre)])
  print()


Year: 1902
Most movies produced in a genre: 1.0 ['Short']

Year: 1912
Most movies produced in a genre: 1.0 ['Biography']

Year: 1913
Most movies produced in a genre: 1.0 ['Comedy' 'Short']

Year: 1915
Most movies produced in a genre: 2.0 ['Drama' 'Comedy']

Year: 1916
Most movies produced in a genre: 1.0 ['Drama']

Year: 1917
Most movies produced in a genre: 1.0 ['Adventure']

Year: 1918
Most movies produced in a genre: 1.0 ['Action' 'Drama' 'Comedy']

Year: 1919
Most movies produced in a genre: 2.0 ['Drama']

Year: 1920
Most movies produced in a genre: 2.0 ['Fantasy']

Year: 1921
Most movies produced in a genre: 2.0 ['Drama']

Year: 1922
Most movies produced in a genre: 2.0 ['Short']

Year: 1923
Most movies produced in a genre: 2.0 ['Comedy']

Year: 1924
Most movies produced in a genre: 1.0 ['Comedy']

Year: 1925
Most movies produced in a genre: 3.0 ['Drama']

Year: 1926
Most movies produced in a genre: 1.0 ['Action' 'Drama' 'Comedy']

Year: 1927
Most movies produced in a genre: 1.0 [